File: 202.快乐数.py
#
# @lc app=leetcode.cn id=202 lang=python3
#
# [202] 快乐数
#

# @lc code=start
class Solution:
    def isHappy(self, n: int) -> bool:
        def step(n):
            res = 0
            while n > 0:
                res += (n % 10) * (n % 10)
                n //= 10
            return res
        
        visited = {}
        while n != 1 and n not in visited:
            visited[n] = True
            n = step(n)
        return n == 1
# @lc code=end

File: 202.快乐数.go
/*
 * @lc app=leetcode.cn id=202 lang=golang
 *
 * [202] 快乐数
 */

// @lc code=start
func isHappy(n int) bool {
	visited := map[int]bool{}
	for n != 1 && !visited[n] { // 当数字重复出现时即进入循环，退出
		visited[n] = true
		n = step(n)
	}
	return n == 1
}

func step(n int) (res int) {
	for n > 0 {
		res += (n % 10) * (n % 10)
		n /= 10
	}
	return
}

// @lc code=end

